vulcan_geometry.py

```diff
@@ -0,0 +1,259 @@
+    corner in the back then increases along the tube upward. The bottom
+    of tube two is in the front.
+
+    back     1 3 5 7
+    front     2 4 6 8
+
+    Setting ``upsidedown=True`` configures the pixels as
+
+    back      1 3 5 7
+    front    2 4 6 8
+
+    This is similar to the incomplete function MantidGeometry.add_double_pack.
+    '''
+    type_element = le.SubElement(instr.root, 'type', name=name)
+    le.SubElement(type_element, 'properties')
+    component = le.SubElement(type_element, 'component', type=tube_type)
+    # z plane is centered in middle of detectors
+    tube_z = (-1. * SEPARATION, SEPARATION)  # 2 rows back, front
+    # x plane is centered between the leftmost and rightmost tubes
+    if upsidedown:
+        x_offset = (-1.25 * SLIP, -1.75 * SLIP)
+    else:
+        x_offset = (-1.75 * SLIP, -1.25 * SLIP)
+    tube_x = np.arange(4, dtype=float) * SLIP   # 4 tube in a row
+    for i, x in enumerate(tube_x):
+        for j, (z, slip) in enumerate(zip(tube_z, x_offset)):
+            le.SubElement(component, 'location', name=f'{name}_{i}_{j}',
+                          x=f'{x + slip:.5f}', z=f'{z:.5f}')
+
+
+# TODO add full orientation
+def addEmptyComponent(instr, type_name: str, rect=None,
+                      x_center: float = 0., z_center: float = 0., rot_y: float = 0.):
+    '''
+    for <location> the rotations happen before the translation
+
+    type_name is reused as idlist name
+    '''
+    component = instr.addComponent(type_name=type_name, idlist=type_name)
+    if rect:
+        rotations = list(rect.euler_rot_yzy)
+        rotations.reverse()  # they are specified in the opposite direction
+        makeLocation(instr, component, None, rect.center, rotations)
+    else:
+        instr.addLocation(component, x=x_center, y=0., z=z_center, rot_y=rot_y)
+
+
+def addBankPosition(instr, bankname: str, componentname: str,
+                    num_panels: int):
+    '''
+    Add all of the 8-packs relative to the bank position
+    '''
+    bank = instr.makeTypeElement(bankname)
+
+    for i in range(num_panels):
+        position_index = i - 0.5*(num_panels-1)
+        packname = 'pack{:02d}'.format(i+1)
+        panel = instr.addComponent(type_name=componentname, root=bank, name=packname)
+        instr.addLocation(panel, x=(SLIP_PANEL * position_index), y=0., z=0., rot_y=0.)
+
+
+def addBankIds(instr, bankname: str, bank_offset: int, num_panels: int):
+    ids = []
+    for i in range(num_panels):
+        panel_offset = bank_offset + PIXELS_PER_PANEL * i
+        ids.extend([panel_offset, panel_offset + PIXELS_PER_EIGHTPACK - 1, None])
+    instr.addDetectorIds(bankname, ids)
+
+
+if __name__ == "__main__":
+    inst_name = "VULCAN"
+    xml_outfile = inst_name+"_Definition.xml"
+    authors = ["Peter Peterson"]
+
+    # boiler plate stuff
+    instr = MantidGeom(inst_name,
+                       comment="Created by " + ", ".join(authors),
+                       valid_from="2020-01-01 00:00:01")
```

Comment:
should this `valid_from` be some time in 2021 instead of a time over a year ago?

Reply:
Very good catch!